pymongo/network_layer.py

```diff
@@ -54,6 +54,7 @@
 _UNPACK_HEADER = struct.Struct("<iiii").unpack
 _UNPACK_COMPRESSION_HEADER = struct.Struct("<iiB").unpack
 _POLL_TIMEOUT = 0.5
+_SSL_RECORD_SIZE = 0x4000
```

Comment:
I'm not sure about this change yet. I'll back it out into a new ticket. Here's the current behavior with stdlib ssl went we send() the entire buffer:
```
test/asynchronous/test_collection.py::AsyncTestCollection::test_large_limit _async_sendall_ssl: <ssl.SSLSocket fd=19, family=2, type=1, proto=0, laddr=('127.0.0.1', 62474), raddr=('127.0.0.1', 27017)>, to_send: 20074138
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: send raised: The operation did not complete (write) (_ssl.c:2406)
_async_sendall_ssl: sent: 0/20074138: 0.0%
_async_sendall_ssl: sent: 20074138/20074138: 100.0%
```
So it appears to send nothing many times over until it finally sends all 20MB in one send() call. That is surprising as I thought a single ssl send() call was limited to 16KB (the max TLS record size).

Reply:
The same test with pyopenssl shows that it does limit a single send() call to 16KB:
```
test/asynchronous/test_collection.py::AsyncTestCollection::test_large_limit _async_sendall_ssl: <pymongo.pyopenssl_context._sslConn object at 0x1091438f0>, to_send: 20074138
_async_sendall_ssl: sent: 16384/20074138: 0.1%
_async_sendall_ssl: sent: 32768/20074138: 0.2%
_async_sendall_ssl: sent: 49152/20074138: 0.2%
_async_sendall_ssl: sent: 65536/20074138: 0.3%
_async_sendall_ssl: sent: 81920/20074138: 0.4%
_async_sendall_ssl: sent: 98304/20074138: 0.5%
_async_sendall_ssl: sent: 114688/20074138: 0.6%
_async_sendall_ssl: sent: 131072/20074138: 0.7%
_async_sendall_ssl: sent: 147456/20074138: 0.7%
_async_sendall_ssl: sent: 163840/20074138: 0.8%
_async_sendall_ssl: sent: 180224/20074138: 0.9%
_async_sendall_ssl: sent: 196608/20074138: 1.0%
_async_sendall_ssl: sent: 212992/20074138: 1.1%
_async_sendall_ssl: sent: 229376/20074138: 1.1%
_async_sendall_ssl: sent: 245760/20074138: 1.2%
_async_sendall_ssl: sent: 262144/20074138: 1.3%
_async_sendall_ssl: sent: 278528/20074138: 1.4%
_async_sendall_ssl: sent: 294912/20074138: 1.5%
_async_sendall_ssl: sent: 311296/20074138: 1.6%
_async_sendall_ssl: sent: 327680/20074138: 1.6%
_async_sendall_ssl: sent: 344064/20074138: 1.7%
_async_sendall_ssl: sent: 360448/20074138: 1.8%
_async_sendall_ssl: sent: 376832/20074138: 1.9%
_async_sendall_ssl: sent: 393216/20074138: 2.0%
_async_sendall_ssl: sent: 409600/20074138: 2.0%
_async_sendall_ssl: sent: 425984/20074138: 2.1%
_async_sendall_ssl: sent: 442368/20074138: 2.2%
_async_sendall_ssl: sent: 458752/20074138: 2.3%
_async_sendall_ssl: sent: 475136/20074138: 2.4%
_async_sendall_ssl: sent: 491520/20074138: 2.4%
_async_sendall_ssl: sent: 507904/20074138: 2.5%
_async_sendall_ssl: sent: 524288/20074138: 2.6%
_async_sendall_ssl: sent: 540672/20074138: 2.7%
_async_sendall_ssl: sent: 557056/20074138: 2.8%
_async_sendall_ssl: sent: 573440/20074138: 2.9%
_async_sendall_ssl: sent: 589824/20074138: 2.9%
_async_sendall_ssl: sent: 606208/20074138: 3.0%
_async_sendall_ssl: sent: 622592/20074138: 3.1%
_async_sendall_ssl: sent: 638976/20074138: 3.2%
_async_sendall_ssl: sent: 655360/20074138: 3.3%
_async_sendall_ssl: sent: 671744/20074138: 3.3%
_async_sendall_ssl: sent: 688128/20074138: 3.4%
_async_sendall_ssl: sent: 704512/20074138: 3.5%
_async_sendall_ssl: sent: 720896/20074138: 3.6%
_async_sendall_ssl: sent: 737280/20074138: 3.7%
_async_sendall_ssl: sent: 753664/20074138: 3.8%
_async_sendall_ssl: sent: 770048/20074138: 3.8%
_async_sendall_ssl: sent: 786432/20074138: 3.9%
_async_sendall_ssl: sent: 802816/20074138: 4.0%
_async_sendall_ssl: sent: 819200/20074138: 4.1%
_async_sendall_ssl: send raised: WantWriteError()
...
_async_sendall_ssl: send raised: WantWriteError()
_async_sendall_ssl: sent: 19791872/20074138: 98.59%
_async_sendall_ssl: sent: 19824640/20074138: 98.76%
_async_sendall_ssl: sent: 19841024/20074138: 98.84%
_async_sendall_ssl: sent: 19857408/20074138: 98.92%
_async_sendall_ssl: sent: 19873792/20074138: 99.00%
_async_sendall_ssl: sent: 19890176/20074138: 99.08%
_async_sendall_ssl: sent: 19906560/20074138: 99.17%
_async_sendall_ssl: sent: 19922944/20074138: 99.25%
_async_sendall_ssl: sent: 19939328/20074138: 99.33%
_async_sendall_ssl: sent: 19955712/20074138: 99.41%
_async_sendall_ssl: sent: 19972096/20074138: 99.49%
_async_sendall_ssl: sent: 19988480/20074138: 99.57%
_async_sendall_ssl: sent: 20004864/20074138: 99.65%
_async_sendall_ssl: sent: 20021248/20074138: 99.74%
_async_sendall_ssl: sent: 20037632/20074138: 99.82%
_async_sendall_ssl: sent: 20054016/20074138: 99.90%
_async_sendall_ssl: sent: 20070400/20074138: 99.98%
_async_sendall_ssl: sent: 20074138/20074138: 100.00%
```

Reply:
This behavior also explains why the current broken code mostly works with stdlib ssl and only fails consistently with PyOpenSSL.